# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/serviceworkers/ServiceWorkerPrivate.cpp
# Commit: 675785853d7a
# Full Hash: 675785853d7a40123826e1b47c49d68f87bd4bd0
# Author: Tim Huang <tihuang@mozilla.com>
# Date: 2024-12-17 17:00:49
# Regressor Bug: 1915355
# File Overlap Count: 1
# Description:
#   Bug 1915355 - Part 5: Exempt third-party cookie blocking if the channel is on the exception list. r=cookie-reviewers,bvandersloot,valentin
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D223384
# ==============================================================================

diff -r de6a26d3e3df -r 675785853d7a dom/serviceworkers/ServiceWorkerPrivate.cpp
--- a/dom/serviceworkers/ServiceWorkerPrivate.cpp	Tue Dec 17 08:57:31 2024 +0000
+++ b/dom/serviceworkers/ServiceWorkerPrivate.cpp	Tue Dec 17 08:57:31 2024 +0000
@@ -49,6 +49,7 @@
 #include "mozilla/ipc/PBackgroundChild.h"
 #include "mozilla/ipc/URIUtils.h"
 #include "mozilla/net/CookieJarSettings.h"
+#include "mozilla/net/CookieService.h"
 #include "nsContentUtils.h"
 #include "nsDebug.h"
 #include "nsError.h"
@@ -556,6 +557,7 @@
   Maybe<RFPTarget> overriddenFingerprintingSettings;
   nsCOMPtr<nsIURI> firstPartyURI;
   bool foreignByAncestorContext = false;
+  bool isOn3PCBExceptionList = false;
   if (!principal->OriginAttributesRef().mPartitionKey.IsEmpty()) {
     net::CookieJarSettings::Cast(cookieJarSettings)
         ->SetPartitionKey(principal->OriginAttributesRef().mPartitionKey);
@@ -583,6 +585,12 @@
           overriddenFingerprintingSettingsArg.emplace(
               uint64_t(overriddenFingerprintingSettings.ref()));
         }
+
+        RefPtr<net::CookieService> csSingleton =
+            net::CookieService::GetSingleton();
+        isOn3PCBExceptionList =
+            csSingleton->ThirdPartyCookieBlockingExceptionsRef()
+                .CheckExceptionForURIs(firstPartyURI, uri);
       }
     }
   } else if (!principal->OriginAttributesRef().mFirstPartyDomain.IsEmpty()) {
@@ -609,6 +617,13 @@
               : nsRFPService::GetOverriddenFingerprintingSettingsForURI(
                     uri, nullptr);
 
+      RefPtr<net::CookieService> csSingleton =
+          net::CookieService::GetSingleton();
+      isOn3PCBExceptionList =
+          isThirdParty ? csSingleton->ThirdPartyCookieBlockingExceptionsRef()
+                             .CheckExceptionForURIs(firstPartyURI, uri)
+                       : false;
+
       if (overriddenFingerprintingSettings.isSome()) {
         overriddenFingerprintingSettingsArg.emplace(
             uint64_t(overriddenFingerprintingSettings.ref()));
@@ -726,8 +741,7 @@
       /* referrerInfo */ nullptr,
 
       storageAccess, isThirdPartyContextToTopWindow, shouldResistFingerprinting,
-      overriddenFingerprintingSettingsArg,
-      false /* TODO: We will set the value in a later patch */,
+      overriddenFingerprintingSettingsArg, isOn3PCBExceptionList,
       // Origin trials are associated to a window, so it doesn't make sense on
       // service workers.
       OriginTrials(), std::move(serviceWorkerData), regInfo->AgentClusterId(),